File: pe7.scala
import scala.math._

/*
By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.

What is the 10 001st prime number?
*/

object pe7 {

  def solve = {
    var primes = List[Int](2);
    val ithPrime = 10001
    var currentValue = 3

    while (primes.length < ithPrime) {
      if ((currentValue % 2) != 0) {
        var possible = true
        println("possible")
        for (prime <- primes) {
          if ((currentValue % prime) != 0)
            possible = possible && true
          else
            possible = possible && false
        }
        if (possible) {
          println("Adding " + currentValue)
          primes = primes :+ currentValue
        }
      }
      currentValue += 2
    }
    println(primes.last)
  }

}

File: pe7/pe7.go
package main

/*
By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.

What is the 10 001st prime number?
*/

import "fmt"

func nthPrime(n int) int {
	primes := []int{}
	list := make([]bool, 10000000)
	list[0] = true
	list[1] = true
	for i, isprime := range list {
		if !isprime {
			primes = append(primes, i)
			for j := i * i; j < len(list); j += i {
				list[j] = true
			}
		}
		if len(primes) > n {
			return primes[n-1]
		}
	}
	return -1
}

func main() {
	fmt.Printf("The 10,001st prime is %d\n", nthPrime(10001))
}
